src/libraries/System.Private.CoreLib/src/System/SR.cs

```diff
@@ -53,7 +53,6 @@ internal static string GetResourceString(string resourceKey, string? defaultStri
         private static int _infinitelyRecursingCount;
         private static bool _resourceManagerInited = false;
 
-        [PreserveDependency(".cctor()", "System.Resources.ResourceManager")]
```

Comment:
@vitek-karas, @MichalStrehovsky - I have some questions on this:

1. I'm a little confused at why this was added in the first place. Why would calling `RunClassConstructor` force that a static constructor be preserved on a Type? If the linker decided that all the code in a `.cctor()` was unnecessary and removed the whole thing, calling `RunClassConstructor` would be just fine right? (just like it is fine to call it on `typeof(ResourceReader)` which doesn't appear to have a `.cctor()`).
2. If this would need to be annotated, how would a developer use `DynamicallyAccessedMembersAttribute` to annotate a Type that is eventually passed into `RuntimeHelpers.RunClassConstructor`?

Reply:
1. If the .cctor has any side effects, linker would not know to keep it and the end result would be to lose those side-effects -> bad. So I added the attribute to make sure we keep it.
2. As per https://github.com/dotnet/runtime/issues/36330 linker now handles `RunClassConstructor` as an intrinsic and will preserve the `.cctor` as necessary. So no annotation is necessary. And also none of the annotations we currently have would work anyway.

Reply:
I'm questioning the validity of removing the .cctor if _anything_ on the type is used.  It's not uncommon to use a .cctor for side-effects, e.g. initializing something in the system.  We did/do it ourselves in a variety of cases, e.g.
https://github.com/dotnet/corefx/blob/c74ea81f2ddad07dc05d695d8244fb2ece7e398d/src/Common/src/Interop/Unix/System.Net.Http.Native/Interop.Initialization.cs#L10-L20

Reply:
1. If the .cctor has any side effects, why does the linker think it is safe to remove it?
2. I'm thinking of the scenario like the following: https://github.com/dotnet/wpf/blob/ac9d1b7a6b0ee7c44fd2875a1174b820b3940619/src/Microsoft.DotNet.Wpf/src/Shared/MS/Internal/SecurityHelper.cs#L124-L127

```
    internal static class SecurityHelper
    {
        internal static void RunClassConstructor(Type t)
        {
            System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor(t.TypeHandle);
        }
```

How could the developers of the above code annotate the `Type` that is passed in to `SecurityHelper.RunClassConstructor`?

Reply:
1. I don't know the answer - I'll have to look into that (I don't remember the rules which linker uses to remove .cctors)
2. We don't have a way to annotate that using the data flow - linker will issue a warning. If I know what the type could be, then I can add the `DynamicDependencyAttribute` and suppress the warning. If I don't, it's simply a pattern we don't have a solution for (just like 10s of other reflection patterns) - maybe "yet"?

Reply:
I wrote the answer to 1 here: https://github.com/dotnet/runtime/pull/36277#discussion_r424261042. It wasn't really needed (we call RunClassConstructor not because of some side effects, but just to frontload it in case there's a problem with running it). If the linker decided to remove the cctor, it would be fine for this to be a no-op. There are legit situations when linker can remove the cctor when it knows it won't be triggered (a beforefieldinit cctor when none of the static fields are accessed is provably not needed because it won't trigger, except for these reflection-like APIs that we analyze/warn about).

2. The cctor would be kept if constructors are kept, so we do have an annotation for this. Cctors are returned from the `Type.GetConstructors` APIs. What we don't have is a way to annotate `RuntimeTypeHandle`. We could add it, but this feels like a niche. The uses in WPF are hopeless for dataflow analysis to analyze anyway (they run on stuff read from XAML).

Reply:
The part of (1) that I'm still confused about is why calling `RuntimeHelpers.RunClassConstructor` would force the preservation of the `.cctor`? If the `.cctor` can be trimmed, or if it must be preserved, is a decision that should be made irregardless of someone calling `RuntimeHelpers.RunClassConstructor` on the Type.

Reply:
```csharp
class Program
{
    static int X = InitX();

    static int InitX()
    {
        Console.WriteLine("Initting X");
        return 42;
    }
    public Program() => Console.WriteLine("In constructor");

    static void Main()
    {
        // Notice we're already running a static method and the cctor didn't trigger yet
        Console.WriteLine("In static method");

        // newobj and still the cctor didn't run
        _ = new Program();

        // Now it runs
        RuntimeHelpers.RunClassConstructor(typeof(Program).TypeHandle);
    }
}
```

The above program prints:

```
In static method
In constructor
Initting X
```

If linker didn't analyze `RunClassConstructor`, the above program would not print "Initting X". That's a difference in behavior between linker/unlinked app and undesirable. In _this specific case_ of the `ResourceManager` class, we would be fine if the call did nothing because we didn't call `RunClassConstructor` for it's side effects: we only called it to make sure that we don't run extra code when we're in the critical path.

`RunClassConstructor` is a reflection-like API, so we just treat is as reflection.

Reply:
OK, that example clears it up for me. Thanks for the explanation.